### PR TITLE
Fix layout with respect to table cells

### PR DESCRIPTION
**Pull request by majvan**
The table cells physically split the textboxes.
The layout scheme is not checking the textbox boundaries
with respect to cells in table and merges textboxes
accross cells.

This commit should improve the work with PDFs with tables.

Signed-off-by: Juraj Vanco <majvan@gmail.com>

### CODE REVIEW

**Comment by RolandColored** [comment]
Can you add an example PDF for this use case, so we can understand the desired effect more clearly? Bonus: Add a test for this PDF.

**Comment by jacobefaust** [comment]
Sorry for my ignorance in advance, but I have what appears to me to be a dumb question.

Does laparams need to have .cell_margin added to it as well?

File: pdfminer/layout.py [hunk 1]
```diff
@@ -582,6 +620,11 @@ def group_textlines(self, laparams, lines):
             if line not in neighbors: continue
             members = []
             for obj1 in neighbors:
+                if line is not obj1:
+                    if self.vsplitted(line, obj1, splitobjs, laparams.cell_margin):
```
Comment (pietermarsman) [inline pdfminer/layout.py]: `LAParams` does not have an attribute `cell_margin` or `split_tables`.

File: pdfminer/layout.py [hunk 2]
```diff
@@ -500,8 +500,46 @@ def __init__(self, bbox):
         self.groups = None
         return
 
+    #detect if between obj0 and obj1 there is a horizontal split line
+    def hsplitted(self, obj0, obj1, splitobjs, cell_margin):
+        cross = False
+        for r in splitobjs:
+            if (r.x0 <= min(obj0.x0, obj1.x0)) and (r.x1 >= max(obj0.x1, obj1.x1)):
+                if   ((r.y0 + cell_margin) >= obj0.y0) and ((r.y0 - cell_margin) <= obj1.y1):
+                    cross = True
```
Comment (pietermarsman) [inline pdfminer/layout.py]: Replace 

```python
cross = True
break
```

by 

```python
return True
```

File: pdfminer/layout.py [hunk 3]
```diff
@@ -500,8 +500,46 @@ def __init__(self, bbox):
         self.groups = None
         return
 
+    #detect if between obj0 and obj1 there is a horizontal split line
```
Comment (pietermarsman) [inline pdfminer/layout.py]: Add this comment as a docstring to the method

File: pdfminer/layout.py [hunk 4]
```diff
@@ -675,6 +718,10 @@ def analyze(self, laparams):
         # textobjs is a list of LTChar objects, i.e.
         # it has all the individual characters in the page.
         (textobjs, otherobjs) = fsplit(lambda obj: isinstance(obj, LTChar), self)
+        if laparams.split_tables:
+            (splitobjs, otherobjs) = fsplit(lambda obj: isinstance(obj, LTRect), self)
```
Comment (pietermarsman) [inline pdfminer/layout.py]: Why do you only select `LTRect`? Are all tables composed of rectangles? Are there other literals that could indicate that two textboxes belong together?

File: pdfminer/layout.py [hunk 5]
```diff
@@ -500,8 +500,46 @@ def __init__(self, bbox):
...
+                elif ((r.y1 + cell_margin) >= obj1.y0) and ((r.y1 - cell_margin) <= obj0.y1):
+                    cross = True
+                    break
+        return cross
+
+    #detect if between obj0 and obj1 there is a vertical split line
+    def vsplitted(self, obj0, obj1, splitobjs, cell_margin):
+        cross = False
+        for r in splitobjs:
+            if (r.y0 <= min(obj0.y0, obj1.y0)) and (r.y1 >= max(obj0.y1, obj1.y1)):
```
Comment (pietermarsman) [inline pdfminer/layout.py]: Improve the readability of this function by using some well-named variables

**Comment by pietermarsman** [comment]
Closing due to lack of activity, failing test and missing issue.

**Comment by jacobefaust** [comment]
Sorry it has taken me so long to respond (I have had a beautiful addition to my family).

I am attaching an example document with irregular tables.  I have improved my fork (the code now passes Travis.ci checks and works on attached example). LTCurve based objects now can be used as barriers during the Textbox and Textline aggregation.

[Pages_from_04sect3.pdf](https://github.com/pdfminer/pdfminer.six/files/4269846/Pages_from_04sect3.pdf)

I need to have this issue reopened so i may issue a pull request.

Thank you for your time in advance.

**Comment by pietermarsman** [comment]
Closing this again in favor of #379

**Pull request closed without merging**